### PR TITLE
Semantic Versioning based Intelligent Routing

### PR DESCRIPTION
**Pull request by pubudu538**
Fixes https://github.com/wso2/apk/issues/1864

## Automation tests
 - Unit tests 
   > Code coverage information
 - Integration tests
   > Details about the test cases and coverage

## Security checks
 - Followed secure coding standards in http://wso2.com/technical-reports/wso2-secure-engineering-guidelines? yes/no
 - Ran FindSecurityBugs plugin and verified report? yes/no
 - Confirmed that this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets? yes/no

### CODE REVIEW

File: adapter/pkg/semanticversion/semantic_version.go [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+// SemVersion is the struct to store the version components of an API
+type SemVersion struct {
+	Version string
+	Major   int
+	Minor   int
+	Patch   *int
+}
+
+// ValidateAndGetVersionComponents validates version string and extracts version components
+func ValidateAndGetVersionComponents(version string, apiName string) (*SemVersion, error) {
```
Comment (AmaliMatharaarachchi) [inline adapter/pkg/semanticversion/semantic_version.go:37]: can't we do this validation in cr validation level (in validation webhook)?

File: adapter/pkg/semanticversion/semantic_version.go [hunk 2]
```diff
@@ -0,0 +1,109 @@
+/*
+ *  Copyright (c) 2023, WSO2 LLC. (http://www.wso2.org) All Rights Reserved.
```
Comment (AmaliMatharaarachchi) [inline adapter/pkg/semanticversion/semantic_version.go:2]:
```suggestion
 *  Copyright (c) 2024, WSO2 LLC. (http://www.wso2.org) All Rights Reserved.
```

File: adapter/internal/discovery/xds/server.go [hunk 3]
```diff
@@ -613,6 +625,11 @@ func GenerateHashedAPINameVersionIDWithoutVhost(name, version string) string {
 	return generateHashValue(name, version)
 }
 
+// GenerateIdentifierForAPIWithoutVersion generates an identifier unique to the API despite of the version
+func GenerateIdentifierForAPIWithoutVersion(vhost, name string) string {
```
Comment (AmaliMatharaarachchi) [inline adapter/internal/discovery/xds/server.go:629]: we can move it to utils if this is an util func. If not, we can make it a private func.

File: adapter/pkg/semanticversion/semantic_version.go (hunk 1 above)
Comment (pubudu538) [reply adapter/pkg/semanticversion/semantic_version.go:37]: For Choreo there are APIs that don't match with this validation. Hence we have to support both API creations. That's why we didnt include this validation in the webhook.

File: adapter/pkg/semanticversion/semantic_version.go (hunk 2 above)
Comment (pubudu538) [reply adapter/pkg/semanticversion/semantic_version.go]: Fixed.

File: adapter/internal/discovery/xds/server.go (hunk 3 above)
Comment (pubudu538) [reply adapter/internal/discovery/xds/server.go]: Moved to a private function.

File: adapter/pkg/semanticversion/semantic_version.go (hunk 1 above)
Comment (AmaliMatharaarachchi) [reply adapter/pkg/semanticversion/semantic_version.go:37]: In there also we can check the config value and perform the validation right?

**Review by AmaliMatharaarachchi — APPROVED** [review]

**Pull request merged**